### PR TITLE
Des 61 filter by pending

### PR DESCRIPTION
**Pull request by bogdan-zaharia**
## Link to JIRA ticket

https://hackney.atlassian.net/jira/software/projects/DES/boards/45?selectedIssue=DES-61

### *What changes have we introduced*

Added request object for getting evidence requests and refactored gateway function to accept this argument
Added evidence request attribute for the state based on document submissions state

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs [hunk 1]
```diff
@@ -1,21 +1,66 @@
...
+        [Test]
+        public void EvidenceRequestHaveCorrectAttributes()
+        {
+            var id = Guid.NewGuid();
+            var createdAt = DateTime.Now;
+            var residentId = Guid.NewGuid();
+            var deliveryMethods = new List<DeliveryMethod>();
+            deliveryMethods.Add(DeliveryMethod.Sms);
+            var documentTypes = new List<string>();
+            documentTypes.Add("passport");
```
Comment (samtgarson) [inline EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs:22]:
```suggestion
            var documentTypes = new List<string>() { "passport" }
```

File: EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs [hunk 2]
```diff
@@ -1,21 +1,66 @@
...
+            var serviceRequestedBy = "some-service";
+            var userRequestedBy = "some-user";
+            var communications = new List<Communication>();
+            communications.Add(TestDataHelper.Communication());
```
Comment (samtgarson) [inline EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs:26]: Same here ☝️

File: EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs [hunk 3]
```diff
@@ -1,21 +1,66 @@
...
+            var documentSubmissions = new List<DocumentSubmission>();
+            documentSubmissions.Add(TestDataHelper.DocumentSubmission());
```
Comment (samtgarson) [inline EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs:28]: Same here ☝️

File: EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs [hunk 4]
```diff
@@ -1,21 +1,66 @@
...
+            evidenceRequest.Communications.Should().BeEquivalentTo(communications);
+            evidenceRequest.DocumentSubmissions.Should().BeEquivalentTo(documentSubmissions);
+        }
+
+        [Test]
+        public static void CanParseDeliveryMethods()
         {
-            _classUnderTest = _fixture.Create<EvidenceRequest>();
+            var deliveryMethods = new List<string>();
+            deliveryMethods.Add("Sms");
```
Comment (samtgarson) [inline EvidenceApi.Tests/V1/Domain/EvidenceRequestTests.cs:59]: same here ☝️

File: EvidenceApi.Tests/V1/Gateways/EvidenceGatewayTests.cs [hunk 5]
```diff
@@ -185,42 +186,71 @@ public void FindReturnsNullWhenADocumentSubmissionWithIdCannotBeFound()
...
+                ServiceRequestedBy = "invalid-service"
+            };
 
-            var result = _classUnderTest.GetEvidenceRequests(serviceRequestedBy);
+            var result = _classUnderTest.GetEvidenceRequests(request);
             result.Should().BeEmpty();
         }
 
-        public List<EvidenceRequest> ExpectedEvidenceRequests(Resident? resident = null)
+        public List<EvidenceRequest> ExpectedEvidenceRequests(Resident? resident = null, EvidenceRequestState? state = null)
```
Comment (samtgarson) [inline EvidenceApi.Tests/V1/Gateways/EvidenceGatewayTests.cs:253]: Not quite sure I understand what this method is doing? It adds all the same attributes to both evidence requests, so how is it testing the filtering functionality

File: EvidenceApi/V1/Domain/EvidenceRequest.cs [hunk 6]
```diff
@@ -39,5 +43,32 @@ public virtual List<DeliveryMethod> DeliveryMethods
...
+        //         if (DocumentTypes.ToArray().All(dt =>
+        //             DocumentSubmissions.Any(ds =>
+        //             ds.State == SubmissionState.Uploaded && ds.DocumentTypeId == dt)
+        //         ))
+        //         {
+        //             return EvidenceRequestState.ForReview;
+        //         }
+        //     }
+        //     return EvidenceRequestState.Pending;
+        // }
```
Comment (samtgarson) [inline EvidenceApi/V1/Domain/EvidenceRequest.cs:72]: 👀

File: EvidenceApi/V1/Gateways/EvidenceGateway.cs [hunk 7]
```diff
@@ -47,16 +46,14 @@ public EvidenceRequest FindEvidenceRequest(Guid id)
...
-                    .Where(x => x.ServiceRequestedBy.Contains(serviceRequestedBy)).ToList();
-            }
             return _databaseContext.EvidenceRequests
-                .Where(x => x.ServiceRequestedBy.Contains(serviceRequestedBy) &&
-                    x.ResidentId.Equals(residentId)).ToList();
+                .Where(x =>
+                    x.ServiceRequestedBy.Contains(request.ServiceRequestedBy) &&
+                    (request.ResidentId == null || x.ResidentId.Equals(request.ResidentId)) &&
+                    (request.State == null || x.State.Equals(request.State))
+                ).ToList();
```
Comment (samtgarson) [inline EvidenceApi/V1/Gateways/EvidenceGateway.cs:56]: This is nice now 👏

File: EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            }
+
+            if (evidenceRequest.DocumentTypes.ToArray().All(dt =>
+                evidenceRequest.DocumentSubmissions.Any(ds =>
+                ds.State == SubmissionState.Approved && ds.DocumentTypeId == dt)
+            ))
+            {
+                return EvidenceRequestState.Approved;
+            }
+            else
```
Comment (samtgarson) [inline EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs:36]: don't need an `else` if there's a `return` above

File: EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        private readonly IEvidenceGateway _evidenceGateway;
+        private readonly IDocumentTypeGateway _documentTypeGateway;
+
+        public UpdateEvidenceRequestStateUseCase(IEvidenceGateway evidenceGateway, IDocumentTypeGateway documentTypeGateway)
+        {
+            _evidenceGateway = evidenceGateway;
+            _documentTypeGateway = documentTypeGateway;
+        }
+
+        public EvidenceRequestState Execute(Guid id)
```
Comment (samtgarson) [inline EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs:21]: This function could do with a refactor/optimisation, but for now its good 👍

**Review by samtgarson — CHANGES_REQUESTED** [review]
Good stuff overall, a couple of comments, but one major one:

I believe the logic in the `UpdateEvidenceRequestState` use case is incorrect, and it's untested.

I'd recommend going back now and starting by writing test cases for the different cases (`Approved`, `ForReview`, `Pending`), and then use them to build the correct logic in the use case

File: EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+        public EvidenceRequest Execute(Guid id)
+        {
+            var evidenceRequest = _evidenceGateway.FindEvidenceRequest(id);
+            if (evidenceRequest == null)
+            {
+                throw new NotFoundException($"Cannot find an evidence request with ID: {id}");
+            }
+
+            if (evidenceRequest.DocumentTypes.ToArray().All(dt =>
```
Comment (neilmendum) [inline EvidenceApi/V1/UseCase/UpdateEvidenceRequestStateUseCase.cs:29]: One way of making this, and the other statement below checking if any evidence requests are in _Uploaded_ state, a bit more readable is to extract the logic into 2 private methods. It's good practice to name them something descriptive, for example `allEvidenceRequestsAreApproved`. See [here](https://refactoring.guru/decompose-conditional) for more info

File: EvidenceApi/V1/UseCase/UpdateDocumentSubmissionStateUseCase.cs [hunk 11]
```diff
@@ -29,6 +35,11 @@ public DocumentSubmissionResponse Execute(Guid id, DocumentSubmissionRequest req
                 throw new NotFoundException($"Cannot find document submission with id: {id}");
             }
 
+            if (String.IsNullOrEmpty(request.State))
```
Comment (neilmendum) [inline EvidenceApi/V1/UseCase/UpdateDocumentSubmissionStateUseCase.cs:38]: Would be good to write a test case for this, unless I missed it

**Review by neilmendum — APPROVED** [review]

**Review by samtgarson — APPROVED** [review]

**Pull request merged**